### ISSUE TITLE
unregistering then re-registering a model doesn't clear the factoryCache

### ISSUE THREAD

**Issue by Leooo** [emberjs/ember.js#15268]
In a unit test:

```javascript
import ServicesProductHoldingBookEngineerMixin from 'ember-commons/models/services-product-holding-book-engineer-mixin';
moduleForModel('appointment', {
  needs: [/*...*/],
  beforeEach() {
    let store = this.store();
    let servicesProductHoldingModel = store.modelFor('servicesProductHolding');
    let servicesProductHoldingModel2 = servicesProductHoldingModel.extend(ServicesProductHoldingBookEngineerMixin);
    getOwner(store).unregister('model:services-product-holding');
    getOwner(store).register('model:services-product-holding', servicesProductHoldingModel2);
    let sph = store.createRecord('servicesProductHolding');//not extended
    console.log('test mixin', sph.get('productDisplayName'));//'productDisplayName' is a property of the mixin, so I expect to see it - except not.
  }
});
```

Part of the solution is to manually delete the factoryCache:

```javascript
let container=getOwner(this).__container__;
delete container.factoryCache['model:services-product-holding']
```

but then the relationships of the previously registered model are not properly cleared: 
```You defined the 'faults' relationship on dummy@model:services-product-holding:, but multiple possible inverse relationships of type dummy@model:services-product-holding: were found on dummy@model:fault```
UPDATE: this is due in a bug on Ember-Data 2.10 in https://github.com/emberjs/data/blob/v2.10.0/addon/-private/system/relationships/ext.js#L311 not linked to this issue so deleting the container factoryCache is above is indeed the right solution.

So basically https://github.com/emberjs/ember.js/issues/11173 is current again

**Comment by stefanpenner** [comment]
I suspect on master the cache that needs to be cleared is named differently https://github.com/emberjs/ember.js/blob/master/packages/container/lib/container.js#L35

**Comment by stefanpenner** [comment]
tests introduced by 11173: https://github.com/emberjs/ember.js/blob/aeeddf7c00b42a340494465252481b5414d87009/packages/ember-application/tests/system/engine_instance_test.js#L36-L55

**Comment by stefanpenner** [comment]
@bekzod got any cycles to lend a hand on this one?

**Comment by bekzod** [comment]
Sure, will look at

**Comment by bekzod** [comment]
@Leooo which version of ember you were looking at ?

**Comment by Leooo** [comment]
Ember & ED 2.10

**Comment by bekzod** [comment]
@Leooo tried following test https://cl.ly/3e2w2I3U2A0l in ember `2.10.0`, `2.10.2` and `master` it does not fail, it might be ember data issue
I also have manually checked everything work cannot reproduce 
https://cl.ly/0e0z2B2K0j2T/Screen%20Shot%202017-05-23%20at%202.27.36%20AM.jp2
Ember	2.10.2
Ember Data	2.10.0

**Comment by Leooo** [comment]
@bekzod ok strange, leave this open I will try to create a failing test in the next days, probably with Ember Data indeed (in which case it will go into the the ED repo I suppose).

Thanks a lot for checking,

L

**Comment by Leooo** [comment]
@bekzod added failing test in https://github.com/emberjs/data/pull/4999

**Comment by rwjblue** [comment]
While poking at this I beefed up some of Ember's tests around `owner.unregister` [here](https://github.com/emberjs/ember.js/pull/15300).  It seems that all is working well on the Ember side, so I'm going to close this and we can dig into the Ember Data issue over in https://github.com/emberjs/data/pull/4999 and https://github.com/emberjs/data/issues/5000.

**Comment by runspired** [comment]
I followed up on this here: https://github.com/emberjs/data/issues/5000#issuecomment-383337040
